File: backend/app/schemas/incident.py
"""Incident API schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class IncidentCreateRequest(BaseModel):
    project_id: UUID
    pipeline_run_id: UUID | None = None
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    source: str = Field(min_length=1, max_length=40)
    severity: str
    priority: str | None = None
    environment: str | None = Field(default=None, max_length=50)


class IncidentUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    severity: str | None = None
    priority: str | None = None
    tags: list[str] | None = None


class IncidentStatusChangeRequest(BaseModel):
    status: str
    comment: str | None = None


class IncidentAssignRequest(BaseModel):
    user_id: UUID
    reason: str | None = None


class ProjectSummary(BaseModel):
    id: UUID
    name: str
    key: str | None = None


class PipelineRunSummary(BaseModel):
    id: UUID
    external_run_id: str | None = None
    provider: str | None = None
    workflow_name: str | None = None
    source_url: str | None = None


class AssigneeSummary(BaseModel):
    id: UUID
    email: str
    full_name: str


class LatestAnalysisSummary(BaseModel):
    id: UUID
    status: str
    classification: dict[str, Any] | None = None
    root_cause_summary: str | None = None


class IncidentListItem(BaseModel):
    id: UUID
    incident_number: str
    title: str
    project: ProjectSummary
    severity: str
    status: str
    environment: str | None = None
    predicted_category: str | None = None
    ai_confidence: float | None = None
    detected_at: datetime
    current_assignee: AssigneeSummary | None = None


class IncidentResponse(BaseModel):
    id: UUID
    incident_number: str
    project_id: UUID
    pipeline_run_id: UUID | None = None
    title: str
    description: str | None = None
    source: str
    status: str
    severity: str
    priority: str | None = None
    environment: str | None = None
    detected_at: datetime
    acknowledged_at: datetime | None = None
    resolved_at: datetime | None = None
    closed_at: datetime | None = None
    created_at: datetime | None = None
    tags: list[str] | None = None


class IncidentDetailResponse(IncidentResponse):
    project: ProjectSummary
    pipeline_run: PipelineRunSummary | None = None
    latest_analysis: LatestAnalysisSummary | None = None
    current_assignee: AssigneeSummary | None = None


class AcknowledgeResponse(BaseModel):
    success: bool = True
    acknowledged_at: datetime


class TimelineEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    event_type: str
    actor_type: str
    title: str
    description: str | None = None
    occurred_at: datetime

File: frontend/src/types/incident.ts
export type IncidentStatus =
  | "detected"
  | "analysing"
  | "open"
  | "in_progress"
  | "resolved"
  | "closed"
  | "analysis_failed"
  | "ignored"
  | "false_positive"
  | "reopened";

export type IncidentSeverity = "critical" | "high" | "medium" | "low";
export type IncidentPriority = "urgent" | "high" | "normal" | "low";

export interface IncidentCreateRequest {
  project_id: string;
  pipeline_run_id?: string | null;
  title: string;
  description?: string | null;
  source: string;
  severity: IncidentSeverity;
  priority?: IncidentPriority | null;
  environment?: string | null;
}

export interface IncidentUpdateRequest {
  title?: string;
  severity?: IncidentSeverity;
  priority?: IncidentPriority;
  tags?: string[];
}

export interface IncidentStatusChangeRequest {
  status: IncidentStatus;
  comment?: string | null;
}

export interface ProjectSummary {
  id: string;
  name: string;
  key: string | null;
}

export interface PipelineRunSummary {
  id: string;
  external_run_id: string | null;
  provider: string | null;
  workflow_name: string | null;
  source_url: string | null;
}

export interface AssigneeSummary {
  id: string;
  email: string;
  full_name: string;
}

export interface LatestAnalysisSummary {
  id: string;
  status: string;
  classification: Record<string, unknown> | null;
  root_cause_summary: string | null;
}

export interface IncidentListItem {
  id: string;
  incident_number: string;
  title: string;
  project: ProjectSummary;
  severity: IncidentSeverity;
  status: IncidentStatus;
  environment: string | null;
  predicted_category: string | null;
  ai_confidence: number | null;
  detected_at: string;
  current_assignee: AssigneeSummary | null;
  /** Optional: present when the backend includes it on the list response. */
  source?: string | null;
}

export interface Incident {
  id: string;
  incident_number: string;
  project_id: string;
  pipeline_run_id: string | null;
  title: string;
  description: string | null;
  source: string;
  status: IncidentStatus;
  severity: IncidentSeverity;
  priority: IncidentPriority | null;
  environment: string | null;
  detected_at: string;
  acknowledged_at: string | null;
  resolved_at: string | null;
  closed_at: string | null;
  created_at: string | null;
  tags: string[] | null;
}

export interface IncidentDetail extends Incident {
  project: ProjectSummary;
  pipeline_run: PipelineRunSummary | null;
  latest_analysis: LatestAnalysisSummary | null;
  current_assignee: AssigneeSummary | null;
}

export interface TimelineEvent {
  id: string;
  event_type: string;
  actor_type: string;
  title: string;
  description: string | null;
  occurred_at: string;
}

export interface IncidentNote {
  id: string;
  incident_id: string;
  author_id: string;
  note_type: string;
  content: string;
  is_pinned: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface NoteCreateRequest {
  note_type: string;
  content: string;
  is_pinned?: boolean;
}

export interface ResolutionSummary {
  id: string;
  resolution_summary: string;
  confirmed_root_cause: string;
  resolution_steps: string[] | null;
  prevention_actions: string[] | null;
  time_spent_minutes: number | null;
  ai_recommendation_used: boolean | null;
  created_at: string | null;
}

export interface ResolveIncidentRequest {
  resolution_summary: string;
  confirmed_root_cause: string;
  resolution_steps: string[];
  prevention_actions: string[];
  time_spent_minutes?: number | null;
  ai_recommendation_used?: boolean | null;
}

export interface ReopenIncidentRequest {
  reason: string;
}

export interface AcknowledgeResponse {
  success: boolean;
  acknowledged_at: string;
}

export interface IncidentAssignRequest {
  user_id: string;
  reason?: string | null;
}
